crates/oxc_span/src/cmp.rs

```diff
@@ -0,0 +1,158 @@
+/// This trait works similarly to [PartialEq] but it gives the liberty of checking the equality of the
+/// content loosely. This would mean the implementor can skip some parts of the content while doing
+/// equality checks.
+/// As an example, In AST types we ignore fields such as [crate::Span].
+///
+/// One should always prefer using the [PartialEq] over this since implementations of this trait
+/// inherently are slower or in the best-case scenario as fast as the [PartialEq] comparison.
+pub trait ContentEq<Rhs: ?Sized = Self> {
+    /// This method tests for contents of `self` and `other` to be equal.
+    #[must_use]
+    fn content_eq(&self, other: &Rhs) -> bool;
+
+    /// This method tests for contents of `self` and `other` not to be equal.
+    /// The default implementation is almost always
+    /// sufficient, and should not be overridden without very good reason.
+    #[inline]
+    #[must_use]
+    fn content_ne(&self, other: &Rhs) -> bool {
+        !self.content_eq(other)
+    }
+}
+
+impl ContentEq for () {
+    #[inline]
+    fn content_eq(&self, _other: &()) -> bool {
+        true
+    }
+
+    #[inline]
+    fn content_ne(&self, _other: &()) -> bool {
+        false
+    }
+}
+
+/// Blanket implementation for references
+impl<A: ?Sized, B: ?Sized> ContentEq<&B> for &A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mutable references
+impl<A: ?Sized, B: ?Sized> ContentEq<&mut B> for &mut A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&mut B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&mut B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mixed references
+impl<A: ?Sized, B: ?Sized> ContentEq<&mut B> for &A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&mut B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&mut B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mixed references
+impl<A: ?Sized, B: ?Sized> ContentEq<&B> for &mut A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
```

Comment:
Are all these impls necessary? Doesn't auto-deref handle this for us without explicit impls? (that's a question, not a statement)

Reply:
Yes Sir! It was a mistake on my part, As it turns out Rust doesn't support auto deref on operators and that's why the standard library has to implement reference blanket implementations for operators. I'll make sure to remove these.

Reply:
These were removed in #5468.